### ISSUE TITLE
Saving HasMany association with already created associated data

### ISSUE THREAD

**Issue by mylux** [cakephp/cakephp#6195]
Hello,
I noticed this behavior in CakePHP 3.0. I am using the dev latest version.

I would like to know if that would be something to be fixed or if this behavior was decided to be the correct one.

When trying to marshall data to entities, the HasMany association is not properly marshalled if the associated data is already created and I just would like to assign them with the entity I am just creating:

Considering we have a Coordinator HasMany Courses and this piece of code:
```
$coordinator_data=[
            'id' => 'on',
            'courses' => [
                [
                    'id' => '8'
                ]
            ]
        ];

$coordinatortoSave=TableRegistry::get('Coordinators')->newEntity($coordinator_data,['associated'=>['Courses'] ]);
debug($coordinatortoSave);
```

I get this result:

```
object(App\Model\Entity\Coordinator) {

	'new' => true,
	'accessible' => [
		'*' => true
	],
	'properties' => [
		'id' => 'on',
		'courses' => [
			(int) 0 => object(App\Model\Entity\Course) {

				'new' => true,
				'accessible' => [
					'name' => true,
					'coordinator_id' => true,
					'submission_begin' => true,
					'submission_end' => true,
					'configuration_id' => true,
					'validity_year' => true,
					'validity_period' => true,
					'appraisers' => true,
					'colleges' => true,
					'students' => true,
					'_joinData' => true,
					'id' => true
				],
				'properties' => [
					'id' => (int) 8
				],
				'dirty' => [
					'id' => true
				],
				'original' => [],
				'virtual' => [],
				'errors' => [
					'name' => [
						'_required' => 'This field is required'
					],
					'configuration_id' => [
						'_required' => 'This field is required'
					]
				],
				'repository' => 'Courses'
			
			}
		]
	],
	'dirty' => [
		'id' => true,
		'courses' => true
	],
	'original' => [],
	'virtual' => [
		(int) 0 => 'name',
		(int) 1 => 'appraiser'
	],
	'errors' => [],
	'repository' => 'Coordinators'

}
```

It means that, even though I am passing the id of an already existent Course in the database, the Marshaller is trying to create a new Course and set its id to 8. Of course I will get a lot of validation errors, because there are mandatory fields not filled, because, obviously, there is no need to fill them in the request once a Course is not being created, but only assigned to a new Coordinator.

Has it already been decided to stay like it is or does this require a bugfix?

**Comment by markstory** [comment]
Were you expecting the marshaller to link the existing hasmany record into the new entity you are creating?

I am not really sure the marshaller should be generating queries and autoloading associated entities for all association types. BelongsToMany is a bit of a special case. With hasOne/hasMany not only would we need to load the existing objects, but also clear/reset foreign keys.

**Comment by mylux** [comment]
Hi Mark. Yes, I do.
I don't think it should be common one resetting a primary key passing it as an associated record in the data structure passed to the marshaller. For me it seems more common to assign an entity by id in this case. If one is interested in reset the primary key, maybe it could be more appropriate passing this primary key to be reset as the main data, not the associated. What do you think?

**Comment by markstory** [comment]
I am concerned with adding more implicit behaviour into the marshaller. Linking to existing belongs to many data is a very common use case, which is why it is  built in. Relinking existing hasMany or hasOne associations seems less frequent. 

Adding support for this will increase complexity and possibly create impossible to resolve situations when nested associations are involved.

**Comment by mylux** [comment]
I catch your point, and I totally agree that linking belongs to many data by id is more frequent than doing the same in a has many association. But, honestly I don't see anything useful when the marshaller builds a new entity with an already existent primary key, because it will result in an error anyway (duplicate primary key). So, IMHO, it seems more useful to load that data than to create new entities with old primary keys.

**Comment by markstory** [comment]
Generally I wouldn't recommend making id accessible for mass assignment. If we were to support linking hasMany objects I would do it using an `_ids` like syntax which prevents the various kinds nested association issues that can come up.

**Comment by mylux** [comment]
Yes, the _ids syntax seems pretty reasonable for me, but this is not implemented with has many associations. If you agree I can start coding something and I will make a PR.

**Comment by brabeji** [comment]
Hello,
I just noticed unexpected behavior that might be related to this (using dev-master). Assuming Posts belongsTo Category - I am trying to patchEntity of
```
$post = [
  'title' => 'foo title',
  'category' => [
    'id' => 1,
    'title' => 'cat 1 title'
  ]
]
```
with
```
$post = [
  'title' => 'foo title edited',
  'category' => [
    'id' => 2
  ]
]
```
there are two different results depending on where accessibility of category id is declared:

in Category entity
```
$post = [
  'title' => 'foo title edited',
  'category' => [
    'id' => 2,
    'title' => 'cat title 1' // this is wierd
  ]
]
```
or by `['associated' => ['Category' => ['accessibleFields => ['id' => true]]]` passed to patchEntity
```
$post = [
  'title' => 'foo title edited',
  'category' => [ // not changed at all
    'id' => 1,
    'title' => 'cat title 1'
  ]
]
```

**Comment by lorenzo** [comment]
@brabeji please don't use already closed tickets for new problems, open a new ticket :)

To answer your question, we do not allow modification of has many entities when they are not already associated to your parent entity. Only association is allowed, not modification.

**Comment by lorenzo** [comment]
btw, I though this ticket was closed, so please ignore me @brabeji

**Comment by brabeji** [comment]
Thank you for your answer wich clears this a bit! Still the major issue is that in second case the accessibility setting is ignored. I will submit separate ticket.

**Comment by brabeji** [comment]
Well, wait a minute.. the actual association I am speaking of is `belongsTo` and I am not modifying anythig :D

**Comment by markstory** [comment]
@lorenzo This issue should have been closed when #6254 was merged.

@brabeji Thanks, a separate issue to address nested `accessibleFields` would be great.
